### PR TITLE
use process groups and kill

### PR DESCRIPTION
**Pull request by steved**
pkill with a parent id doesn't work for single processes (i.e. sleep infinity).
Find out the pgid, and then use kill on that.

This stuff still doesn't work as well as I'd like, let's try a new tactic.

@zendesk/samson

### CODE REVIEW

File: app/models/terminal_executor.rb [hunk 1]
```diff
@@ -36,7 +36,11 @@ def execute!(*commands)
   def stop!(signal)
     # Need pkill because we want all
     # children of the parent process dead
-    system "pkill", "-#{signal}", "-P", pid.to_s if pid
+    if pgid
+      system('kill', "-#{signal}", "-#{pgid}")
+    elsif pid
+      system "pkill", "-#{signal}", "-P", pid.to_s
+    end
```
Comment (jonmoter) [inline app/models/terminal_executor.rb]: Consistent use of parens, please, for those of us with mild OCD tendencies.  :-)

**Comment by jonmoter** [comment]
My knowledge of how PGID works is limited.  Is there any scenario when the Puma process would have the same PGID as the spawned process?  I want to be careful we don't accidentally kill the Samson web server when trying to stop a job.

**Comment by steved** [comment]
PTY#spawn resets the process group: https://github.com/ruby/ruby/blob/trunk/ext/pty/pty.c#L100-L102

**Comment by jonmoter** [comment]
Okay, let's give this a try.  With any luck, this will fix [SAMSON-157](https://zendesk.atlassian.net/browse/SAMSON-157).

:+1:

**Pull request merged**